### ISSUE TITLE
multitrial ft_databrowser

### ISSUE THREAD

**Issue by dnacombo** [fieldtrip/fieldtrip#1968]
**Reviewing multitrial data can be cumbersome**

In cognitive neuroscience experiments, reviewing all data "by hand" is common practice.
For **trial-based experiments**, reviewing single trials is difficult with the current `ft_databrowser`.
In the current `ft_databrowser`, two modes exist for this: continuous, and not continuous.

When `cfg.continuous == 'yes'` the data is shown in the original time frame, with gaps between trials:

![image](https://user-images.githubusercontent.com/4716235/152345687-4f9cc024-89dc-47bd-9aad-6651c7eb6a4a.png)

When `cfg.continuous == 'no'`, one trial only is shown:

![image](https://user-images.githubusercontent.com/4716235/152345926-95ea831b-6919-458b-9691-abfc7557b629.png)

Reviewing hundreds of trials can be really a pain because of this, and because drawing can be slow.

**Alternative solutions**
Displaying more trials on one screen is doable :

1. Tweaking the not continuous mode seems like a real challenge because the single trial data is at the heart of the plotting engine.
2. Tweaking the continuous mode is a workaround which could work, but requires temporarily adapting all time stamps in the dataset.

I went for 2. in this #PR.

**Drawbacks**

- The displayed dataset is shown on a continuous time scale, which does not correspond to trial-based time.
- Added artifact windows may span several trials.
- Consider that sampleinfo is increasing and non overlapping
- Had to comment out ft_warnings in `ft_plot_vector` and `ft_plot_text` probably not acceptable, but a strong slow-down

**Process**

- I wrote a wrapper for `ft_databrowser` called `ft_databrowser_multitrial`
- Added 1 option: `cfg.ntrialswin`: the number of trials on the first screen
- Compute `chunksout` of time to be removed from `sampleinfo`
- First shift all event time samples to account for `chunksout`
- Add blank boundary events at first sample of each trial
- Shift artifacts

**Comment by dnacombo** [comment]
This is how it looks now:

![image](https://user-images.githubusercontent.com/4716235/152355732-42ecb8e7-e38f-4dbd-b45d-0dc4b921e495.png)

**Comment by robertoostenveld** [comment]
hmm, I understand the desire for the functionality, but am quite worried about the side effects that would be created like this. E.g. when people make and mark selections for artifacts, or make selections and right-click to execute a "browsefun".

The worry partially relates to difficult future maintenance, partially from the function getting too large and complex. In other main FT functions we have been using the strategy to split the function into separate private parts (like ft_prepare_mesh, which calls various private prepare_mesh_xxx functions). And in some cases a single main function has been split in multiple. I don't like the architecture of adding a thin wrapper around a "heavy" function. Note that we do have examples, like  ft_clusterplot.m which wraps around ft_topoplotTFR.

Considering the functionality of ft_databrowser, I would say that it can already be split over 
- viewmode_vertical
- viewmode_butterfly
- viewmode_component
and I have been thinking (but never properly realized) viewmode_multiplot, and now you come with viewmode_multitrial. 

Perhaps it is an idea taking a fresh look at the whole function and seeing whether we can refactor it over multiple private functions. If that works, we would create an improved databrowser framework that could then better be extended.

**Comment by dnacombo** [comment]
I see what you mean.

- artfcdef selections happening during `ft_databrowser` call inside `ft_databrowser_artifact` are accounted for (as far as I could [test](https://github.com/fieldtrip/fieldtrip/pull/1969/commits/8105eb7a1157ceb9c7e0068931ec5d57baceb465) for my own purpose today).
- For "browsefun" fonctions, are you referring to the `selfun` and `selcfg` fields of `cfg`? Looking into it in the coming days, I will see how problematic it is. In principle, all fields of the `cfg` structure are passed through. For now, only `event` is restored to original (see below), and `artfctdef.xxx` fields are updated to account for the `chunksout` mentioned above
- After enough testing and checking that all/most use-cases are accounted for, I would try to include this inside `ft_databrowser` for transparent functionality. Perhaps with a condition like `if istrue(cfg.continuous) & numel(data.trial) > 1` and a nested call to `ft_databrowser`. This isn't totally clean coding practice but I think it could be done properly...
- Still worried with current implementation about overlapping trials, but should be easy to fix.
- Also I see that for `selectmode`, I should account recompute `event` samples. So far I consider that `event` doesn't change inside `ft_databrowser` and I just return the input structure. Now that i have an algorithm for events and artifacts. I don't anticipate a big problem. Let me know if that is useful.
- We're independent of the `viewmode` here, as far as i understand. Tried with this functionality and it seems to work independently #1970.
- Revamping the whole function would require consulting. I'm not aware of all usecases.

**Comment by robertoostenveld** [comment]
> We're independent of the viewmode here, as far as i understand.

No, you _implemented_ (or hacked) it to be independent of viewmode. But considering the possible future functionality to view as a multiplot per trial, I think it is not desirable to be independent of viewmode. The viewmode will help to determine which options make sense and which not. E.g. consider this

<img width="672" alt="Screenshot 2022-02-04 at 14 19 09" src="https://user-images.githubusercontent.com/899043/152535973-07cd84c8-d408-40e4-b198-1de090f825a4.png">

or this

<img width="672" alt="Screenshot 2022-02-04 at 14 22 57" src="https://user-images.githubusercontent.com/899043/152536277-afaf2fb2-290c-4628-a660-45c3a99d0917.png">

I hope you see that there are other choices possible that would not be possible with a hacked-together dataset in a thin wrapper function.

**Comment by robertoostenveld** [comment]
This https://github.com/robertoostenveld/fieldtrip/tree/databrowser-multiplot is a quick implementation of the multiplot version. Artifacts are not handled correctly; this display would in principle allow for different segments per channel to be selected.

Just to zoom out from implementation details: have you considered using/improving ft_rejectvisual with cfg.method=trial instead?

**Comment by dnacombo** [comment]
> I hope you see that there are other choices possible that would not be possible with a hacked-together dataset in a thin wrapper function.

Not a surprise.
But it works nicely for the use I am aiming for: reviewing trials faster than one by one. It's also a display that people are used to.

> This https://github.com/robertoostenveld/fieldtrip/tree/databrowser-multiplot is a quick implementation of the multiplot version. Artifacts are not handled correctly; this display would in principle allow for different segments per channel to be selected.

I see the idea. Will keep that in mind.

> Just to zoom out from implementation details: have you considered using/improving ft_rejectvisual with cfg.method=trial instead?

Yeah I did play around with it a little bit. Could not get rid of the trial number title on top of each plot and these days I have 1200 trials to review so the window was clogged... (and that below is on a fairly large screen).

![image](https://user-images.githubusercontent.com/4716235/152554578-175a0946-11fd-4156-b44d-0522aebef769.png)

**Comment by robertoostenveld** [comment]
1200 is a lot. That indeed won't work well on a single screen. Or [this](https://architizer.com/blog/practice/tools/the-worlds-widest-computer-monitor/) perhaps?

Did you consider using the cfg.continuous option in ft_redefinetrial? That is relatively new and can be used to stitch trials back together. With a small `sampleinfo`  hack you could then use ft_databrowser on the single continuous segment with cfg.blocksize=5 or so for 5 seconds, or anything you like.

If you combine that with an extra channel that contains the original sample numbers, you can easily map the visual artifacts in the pseudo-continuous data back onto the original samples.

The whole script would become something like 

```
% add the sample indices as extra channel
data.label(end+1) = {'origsample'};
for i=1:numel(data.trial)
  data.trial{i}(end+1,:) = data.sampleinfo(i,1):data.sampleinfo(i,2);
end

begsample = [0 cumsum(cellfun(@length, data.trial(1:end-1)))]' + 1;
endsample = cumsum(cellfun(@length, data.trial))';

% make it continuous
data.sampleinfo = [begsample endsample];

% stitch together
cfg = [];
cfg.continuous = 'yes';
data_pseudocontinuous = ft_redefinetrial(cfg, data) 

cfg = [];
cfg.plotevents = 'no';
cfg.viewmode = 'vertical';
cfg.continuous = 'yes';
cfg.channel = {'all', 'origsample'};
cfg = ft_databrowser(cfg, data_pseudocontinuous)

% map the pseudocontinuous sample numbers back on the original sample numbers
origsample = data_pseudocontinuous{1}(end,:) % this is also stitched over the trials
origartifact = origsample(cfg.artfctdef.visual.artifact);
```

I think this is also largely what you are achieving with #1969, but then in (almost) clean native FieldTrip code.

**Comment by dnacombo** [comment]
I see. Nice and simple. But am I wrong thinking that original events and artifacts are not displayed in databrowser in this way?

Also, I pass trial-based data to ft_databrowser with cfg.continuous = 'yes', and it works. So turning it to pseudo_continuous doesn't seem to be necessary.

In #1969, I did forward them through. It's convenient during reviewing to see the original events, and possibly precomputed artifact definitions.

New commit to #1969 uses sample indices as you did above to switch back and forth between time frames.

So if I recap:

1. update sampleinfo so that trials are on a continuous time line
2. update events and artifacts so they show up in the new time line
3. pass to ft_databrowser
4. realign artfctdef out of ft_databrowser in the original time line
5. return that in an updated cfg

Still not sure how this will handle trials overlapping in time.

**Comment by dnacombo** [comment]
hm actually, I have an issue with artfcts that fall outside of trials.
They are not in the `origsample` virtual channel so I loose them on the way.

Need to revert to the previous version.

To be clear, I just explored 2 ways of virtually dropping time segments between trials so as to be able to stitch trials together in databrowser:

1. subtracting time samples of cropped segments iteratively.
2. keeping track of original time samples in a virtual channel.

Problem with 1: none that I can figure out at the moment (need to test overlapping trials)

Problem with 2: event and artfct samples outside of trials are lost. Also this is slower because of lots of `find`.

**Comment by robertoostenveld** [comment]
> But am I wrong thinking that original events and artifacts are not displayed in databrowser in this way?

No, you are right. The script could be extended to modify those as well. In any case, I would work with a channel (or vector) that contains the sample numbers. Or possibly two channels: one with the original sample numbers and one with the shifted sample numbers. Then those two can be used to map back and forth each data object between two representations.

However, I think that the use-case is not sufficiently mature to consider adding a function to FieldTrip. I think that better at this moment would be to document some use-cases where default ft_databrowser is suboptimal and explain how to work around those in the script outside FT. Possible cases (not mutually exclusive): 

1. I have 1200 trials and don't want to click 1200 times
2. I have trials in memory that overlap
3. I have trials with a large ITI that I want to plot as continuous, but without the large gaps in between

rather than continuing with #1969 I therefore suggest to make an example page on the web site (or possibly start with a gist to discuss it).

**Comment by dnacombo** [comment]
> I therefore suggest to make an example page on the web site (or possibly start with a gist to discuss it)

Can you point me in the right direction? do you have an example I could start from?

**Comment by robertoostenveld** [comment]
Examples are preferably executable by people that want to learn from them, therefore I suggest you start with an EEG dataset from our FTP server. Using that EEG file, you can hand-craft a `cfg.trl` structure with the features that you are struggling with. So you simply ignore the actual triggers, but just cut the file in 1200 trials, with overlap between the trials, or with large spacing between the trials. Once that preprocessed data structure can be reproduced (also by me), we can document the various opportunities too deal with artifact removal in that dataset.

You can start a gist by clicking in the upper right corner in github, next to your profile picture. [Example pages](https://github.com/fieldtrip/website/tree/master/example) are simply markdown files with documentation/explanation and inline code.

**Comment by robertoostenveld** [comment]
I just looked on the ftp, we could also add a directory `pub/fieldtrip/example/databrowser` and put one of your actual files in there. That would make it even more realistic.
